ibis/omniscidb/operations.py

```diff
@@ -313,6 +313,123 @@ def extract_field_formatter(translator, expr, sql_attr=sql_attr):
     return extract_field_formatter
 
 
+def _n_weeks_year(year: ir.NumericValue):
+    """
+    Return the total number of weeks for the given year.
+
+    According to ISO 8601, a year can have 52 or 53 weeks.
+
+    The long years, with 53 weeks in them, occurs when:
+
+    * the year is a common year, starting on Thursday
+    * the year is a leap year, starting on Wednesday
+    * the year ends on Thursday
+    * the year is a leap year, ending on Friday
+    * years in which 1 January or 31 December are Thursdays
+
+    Otherwise, the year has 52 weeks.
+
+    Note: common year has 365 days and leap year has 366 days.
+
+    Parameters
+    ----------
+    year : ir.NumericValue
+
+    Returns
+    -------
+    ir.Expr
+
+    See Also
+    --------
+    https://en.wikipedia.org/wiki/ISO_week_date
+    """
+    # adjustment factor for week year extraction
+    def year_group(year: ir.NumericValue):
+        """
+        Return the year group cycle.
+
+        This group helps to check if it is 52 or 53 weeks.
+        For example, the year has 53 weeks if the year group is 4 or if
+        the group of the previous year is 3 (when the given year is leap).
+
+        Parameters
+        ----------
+        year : ir.NumericValue
+
+        Returns
+        -------
+        ir.Expr
+          Return a value between 0 and 6
+
+        See Also
+        --------
+        https://en.wikipedia.org/wiki/ISO_week_date
+
+        """
+        return (year + (year // 4) - (year // 100) + (year // 400)) % 7
+
+    needs_adjustment = (year_group(year) == 4) | (year_group(year - 1) == 3)
+
+    return 52 + (ibis.case().when(needs_adjustment, 1).else_(0).end())
+
+
+def _week_of_year_preliminary(
+    date_expr: Union[ir.DateValue, ir.TimestampValue]
+) -> ir.Expr:
+    """
+    Return a preliminary week of year (without adjustments).
+
+    Parameters
+    ----------
+    date_expr : ibis.expr.types.DateValue or ibis.expr.types.TimestampValue
+
+    Returns
+    -------
+    ir.Expr
+      Possible values between 0 and 53
+
+    See Also
+    --------
+    https://en.wikipedia.org/wiki/ISO_week_date
+    """
+    result = date_expr.day_of_year() - date_expr.day_of_week.index() + 11
+    return (result // 7).cast('int16')
+
+
+def _extract_week_of_year(translator, expr) -> str:
+    """
+    Extract the week of the year expression for a given date/datetime.
+
+    It uses the value of the preliminary week of the year for the given date
+    expression and:
+
+    1) if it is less than 1, it returns the total weeks of the previous year
+       for the given date (last week for the previous year).
+    2) if it is greater than the total weeks of the year for the given date,
+       it returns 1 (first week for the next year)
+
+    See Also
+    --------
+    https://en.wikipedia.org/wiki/ISO_week_date
+    """
+    date = expr.op().args[0]
+
+    week = _week_of_year_preliminary(date)
```

Comment:
Creating a separate function for `_week_of_year_preliminary` seems unnecessary and overcomplicated:
```suggestion
    week_of_year_without_adjustment = ((date.day_of_year() - date.day_of_week.index() + 11) // 7).cast('int16')
```

Reply:
thanks @datapythonista I did this change. just keep the variable name to `week` and added a comment about this there.